Convert Swift Differentiable Programming Implementation Overview to Markdown

philipturner opened the pull request:
<!-- What's in this pull request? -->
@dan-zheng requested that I make this pull request to [apple/swift](https://github.com/apple/swift). I converted a file linked as Swift documentation to Markdown. This is currently a Google Doc, which is less appropriate to collaborate on this setting, and has a pending request to turn it into Markdown. 

I did the following in this PR:
- Created a new file called `DifferentiableProgrammingImplementation.md` in `docs`.
- Created a `DifferentiableProgrammingImplementation` directory in `docs/assets` and added two images.
- Redirected the bit.ly link in the [Differentiable Programming Manifesto](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md) to the new Markdown file.

For the changes I made while converting from the Google Doc, see the [original pull request](https://github.com/tensorflow/swift-apis/pull/1187) to tensorflow/swift-apis.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

philipturner committed:
Add files via upload

philipturner committed:
Add files via upload

philipturner committed:
Delete the-swift-compilation-pipeline.png

philipturner committed:
Add files via upload

philipturner committed:
Create DifferentiableProgrammingImplementation

philipturner committed:
Delete swift-compilation-pipeline.png

philipturner committed:
Delete differentiation-control-flow.png

philipturner committed:
Create temp.md

philipturner committed:
Add files via upload

philipturner committed:
Delete temp.md

philipturner committed:
Rename DifferentiableProgrammingImplementation to DifferentiableProgrammingImplementation.md

philipturner committed:
Update DifferentiableProgrammingImplementation.md

philipturner committed:
Update DifferentiableProgramming.md

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+### Visiting the original function to create a JVP
+
+In the Swift compiler, we have a class called [`JVPEmitter`](https://github.com/apple/swift/blob/tensorflow/lib/SILOptimizer/Mandatory/Differentiation.cpp) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate th ...
```
This is out of date with the current state of main.  I think that we should not be referencing anything on the tensorflow branch.

philipturner replied:
Could you replace the link with the correct one then?
```suggestion
In the Swift compiler, we have a class called [`JVPEmitter`](https://github.com/apple/swift/blob/main/lib/SILOptimizer/Mandatory/Differentiation.cpp) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate the JVP function. Taking a look at the `JVPEmitter` class, there are methods like `visitApplyInst`, `visitStructExtractInst`, etc. Each of these methods visit an instruction that is important in the generation of the JVP and differential, and emits a newly mapped version. We handle each type of instruction differently, explained below.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+One important note is that in the JVP, we visit every single instruction in the original function - sometimes it’s an exact copy, and other times there is some special logic we wrote to handle it differently (e.g. like control-flow discussed below). This is so that the JVP function behaves just like the original function. With this, if the original function has a print statement, the JVP will as  ...
+
+In Swift code, the generated differential function can be written as a closure, capturing the “callee differentials” from the JVP. But in SIL, all functions are top-level; closures are represented as top-level functions with captured values as an explicit argument. This requires a differential struct which will be discussed in the next section. What’s important here is that the `JVPEmitter` emits ...
```
Again, this does not match the current state.

```suggestion
In Swift code, the generated differential function can be written as a closure, capturing the "callee differentials" from the JVP. But in SIL, all functions are top-level; closures are represented as top-level functions with captured values as an explicit argument. This requires a differential struct which will be discussed in the next section. What’s important here is that the `JVPEmitter` emits code both in a top level JVP function, but also a corresponding top level differential function.
```

philipturner replied:
What changes must be made to this text then? (aside from the quotation correction)

compnerd replied:
There is no `JVPEmitter`, so what actually generates that?  This largely impacts the last sentence.

philipturner replied:
@dan-zheng could you patch this sentence since you know the most about this?

rxwei replied:
It's been renamed to `JVPCloner`.

philipturner replied:
```suggestion
In Swift code, the generated differential function can be written as a closure, capturing the "callee differentials" from the JVP. But in SIL, all functions are top-level; closures are represented as top-level functions with captured values as an explicit argument. This requires a differential struct which will be discussed in the next section. What’s important here is that the `JVPCloner` emits code both in a top level JVP function, but also a corresponding top level differential function.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+```swift
+// Struct containing differential functions.
+// Partially-applied to `@df_foo` in `@jvp_foo`.
+struct foo_bb0_DF_src_0_wrt_0 {
````
```suggestion
struct f_bb0_DF_src_0_wrt_0 {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+
+```swift
+// Struct containing differential functions.
+// Partially-applied to `@df_foo` in `@jvp_foo`.
````
```suggestion
// Partially-applied to `@df_f` in `@jvp_f`.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+}
+
+// JVP: replaces all function applications with JVP applications.
+sil @jvp_foo : $(Float) -> (Float, (Float) -> Float) {
```
```suggestion
sil @jvp_f : $(Float) -> (Float, (Float) -> Float) {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+  (%y2, %df_cos) = apply @jvp_cos(%x)
+  (%y3, %df_mul) = apply @jvp_mul(%y1, %y2)
+  // Partially-apply to get a differential.
+  %df_struct = struct $foo_bb0_DF_src_0_wrt_0 (%df_sin, %df_cos, %df_mul)
```
```suggestion
  %df_struct = struct $f_bb0_DF_src_0_wrt_0 (%df_sin, %df_cos, %df_mul)
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+  (%y3, %df_mul) = apply @jvp_mul(%y1, %y2)
+  // Partially-apply to get a differential.
+  %df_struct = struct $foo_bb0_DF_src_0_wrt_0 (%df_sin, %df_cos, %df_mul)
+  %df = partial_apply @df_foo(%df_struct)
```
```suggestion
  %df = partial_apply @df_f(%df_struct)
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+
+// Differential: apply differentials to tangent values.
+sil @df_foo : $(Float, foo_bb0_DF_src_0_wrt_0) -> (Float) {
```
```suggestion
sil @df_f : $(Float, f_bb0_DF_src_0_wrt_0) -> (Float) {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+### Differential struct creation
+
+To dive deeper into the differential struct, the reason why we need these “callee differentials” is that not all differentials are linear, so they capture some sort of “state” from the original function call. For example, in multiplication, the differential is `dx * y + dy * x`. In the differential function, we don’t have access to the values of `x` and `y` - only `dx` and `dy`. Additionally, in  ...
```
```suggestion
To dive deeper into the differential struct, the reason why we need these "callee differentials" is that not all differentials are linear, so they capture some sort of "state" from the original function call. For example, in multiplication, the differential is `dx * y + dy * x`. In the differential function, we don’t have access to the values of `x` and `y` - only `dx` and `dy`. Additionally, in the JVP call, the differential function it returned would have copied these values into the function so we only have `dx` and `dy`. So if we called the function with `x = 4`, `y = 5`, the differential would be `5 * dx + 4 * dy`. Thus, we need to pass these callee differentials into the overall/new differential function we generate by taking an additional "differential struct" argument, which represents a bundle of all of the "captured callee differentials". The JVP then constructs an instance of the "differential struct" and [partially applies](https://en.wikipedia.org/wiki/Partial_application) them to the generated differential to get a differential value which it returns as part of the (original, diff) tuple in the JVP.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+These structs consist of two different types of values: the differential function that maps to the original function call which we got from the second element of the JVP call, and also branching enums.
+
+In order to define the differential struct type, we preemptively go over the entire function we are going to differentiate in order to generate the struct. Using the same approach as with `JVPEmitter` when emitting code in the differential, we visit all instructions that we deem are needed to take the derivative of. We then calculate the expected differential type of the function, and add that as ...
```
Please remove the reference to the `JVPEmitter`.

```suggestion
In order to define the differential struct type, we preemptively go over the entire function we are going to differentiate in order to generate the struct. When emitting code in the differential we visit all instructions that we deem are needed to take the derivative of the function. We then calculate the expected differential type of the function, and add that as a field to the struct.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+In order to define the differential struct type, we preemptively go over the entire function we are going to differentiate in order to generate the struct. Using the same approach as with `JVPEmitter` when emitting code in the differential, we visit all instructions that we deem are needed to take the derivative of. We then calculate the expected differential type of the function, and add that as ...
+
+```swift
+func foo(_ x: Float) -> Float {
````
```suggestion
func f(_ x: Float) -> Float {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+  return 2 * x // only active in one result!
+}
+
+struct foo_bb0_DF_src_0_wrt_0 {
```
```suggestion
struct f_bb0_DF_src_0_wrt_0 {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+In order to handle control flow, we need to create a struct for each basic block. In addition to this, we need an enum field that has the successor basic block struct as a payload value on the enum case. The reason this is required is that control flow is dynamic and what route we take down the control flow is determined during runtime. As such, we need to dynamically create instances of these st ...
+
+```swift
+func bar(_ x: Float) -> Float {
````
```suggestion
func m(_ x: Float) -> Float {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+Another additional transformation that is done in the JVP is the creation of trampoline blocks. These are basic blocks which deal with handling the transition from one basic block to another, specifically regarding the differential structs. For a concrete example, we would roughly have the following SIL code for the control flow function above:
+
+> NOTE: The SIL code below is simplified in both naming and how each instruction looks like. For more information regarding how each instruction looks like in actual code, refer to [SIL.rst](https://github.com/apple/swift/blob/tensorflow/docs/SIL.rst).
```
```suggestion
> NOTE: The SIL code below is simplified in both naming and how each instruction looks like. For more information regarding how each instruction looks like in actual code, refer to [SIL.rst](SIL.rst).
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+> NOTE: The SIL code below is simplified in both naming and how each instruction looks like. For more information regarding how each instruction looks like in actual code, refer to [SIL.rst](https://github.com/apple/swift/blob/tensorflow/docs/SIL.rst).
+
+```swift
+func bar(_ x: Float) -> Float {
````
```suggestion
func m(_ x: Float) -> Float {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+- Here we do the same thing as we have been doing, which is partially applying the overall bb0 struct we created in bb0:
+    - `%bb0_struct = alloc_stack $StructBB0`
+        - This has the nested differential structs for all other relevant differential structs for all other basics blocks, and we partially apply it for the differential of `bar` and return the original result + differential tuple pair
```
```suggestion
        - This has the nested differential structs for all other relevant differential structs for all other basics blocks, and we partially apply it for the differential of `m` and return the original result and differential tuple pair
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+  return (%orig_result, %diff_bar_instance)
+```
+
+- Here we do the same thing as we have been doing, which is partially applying the overall bb0 struct we created in bb0:
````
```suggestion
- Here we do the same thing as we have been doing, which is partially applying the overall `StructBB0` struct we created in basic block 0:
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+## VJP and pullback generation
+
+Similar to generated JVP functions, generated VJP functions are clones of the original function, with function applications replaced with VJP applications. Each VJP application returns the original result of application in addition to a pullback value (named “callee pullback” to distinguish it from the “generated pullback function”).
```
```suggestion
Similar to generated JVP functions, generated VJP functions are clones of the original function, with function applications replaced with VJP applications. Each VJP application returns the original result of application in addition to a pullback value (named "callee pullback" to distinguish it from the "generated pullback function").
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+Similar to generated JVP functions, generated VJP functions are clones of the original function, with function applications replaced with VJP applications. Each VJP application returns the original result of application in addition to a pullback value (named “callee pullback” to distinguish it from the “generated pullback function”).
+
+The generated pullback function takes the partial derivatives with respect to outputs and returns the partial derivative with respect to the input. There is a mapping between _active values in the original function_ and their _adjoint values (i.e. partial derivatives) in the pullback function_. The pullback captures the “callee pullbacks” from the VJP, and replaces _original function applications ...
```
```suggestion
The generated pullback function takes the partial derivatives with respect to outputs and returns the partial derivative with respect to the input. There is a mapping between _active values in the original function_ and their _adjoint values (i.e. partial derivatives) in the pullback function_. The pullback captures the "callee pullbacks" from the VJP, and replaces _original function applications using original values_ with _"callee pullback" applications using adjoint values_.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+
+```swift
+// VJP: replaces all function applications with VJP applications.
+sil @vjp_foo : $(Float) -> (Float, (Float) -> Float) {
````
```suggestion
sil @vjp_f : $(Float) -> (Float, (Float) -> Float) {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+
+// VJP: replaces all function applications with VJP applications.
+sil @vjp_foo : $(Float) -> (Float) -> Float {
```
```suggestion
sil @vjp_f : $(Float) -> (Float) -> Float {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+}
+```
+
+As explained above in the “JVP and differential generation” section, closures do not exist in SIL. A more accurate, closure-free pseudocode looks like:
````
```suggestion
As explained above in the "JVP and differential generation" section, closures do not exist in SIL. A more accurate, closure-free pseudocode looks like:
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+```swift
+// Struct containing pullback functions.
+// Partially-applied to `@pb_foo` in `@vjp_foo`.
+struct foo_bb0_PB_src_0_wrt_0 {
````
```suggestion
struct f_bb0_PB_src_0_wrt_0 {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+
+```swift
+// Struct containing pullback functions.
+// Partially-applied to `@pb_foo` in `@vjp_foo`.
````
```suggestion
// Partially-applied to `@pb_f` in `@vjp_f`.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+}
+
+// VJP: replaces all function applications with VJP applications.
+sil @vjp_foo : $(Float) -> (Float, (Float) -> Float) {
```
```suggestion
sil @vjp_f : $(Float) -> (Float, (Float) -> Float) {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+  (%y2, %pb_cos) = apply @vjp_cos(%x)
+  (%y3, %pb_mul) = apply @vjp_mul(%y1, %y2)
+  // Partially-apply to get a pullback.
+  %pb_struct = struct $foo_bb0_PB_src_0_wrt_0 (%pb_sin, %pb_cos, %pb_mul)
```
```suggestion
  %pb_struct = struct $f_bb0_PB_src_0_wrt_0 (%pb_sin, %pb_cos, %pb_mul)
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+  (%y3, %pb_mul) = apply @vjp_mul(%y1, %y2)
+  // Partially-apply to get a pullback.
+  %pb_struct = struct $foo_bb0_PB_src_0_wrt_0 (%pb_sin, %pb_cos, %pb_mul)
+  %pb = partial_apply @pb_foo(%pb_struct)
```
```suggestion
  %pb = partial_apply @pb_f(%pb_struct)
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+}
+
+// Pullback: apply pullbacks to adjoint values.
+sil @pb_foo : $(Float, foo_bb0_PB_src_0_wrt_0) -> (Float) {
```
```suggestion
sil @pb_f : $(Float, f_bb0_PB_src_0_wrt_0) -> (Float) {
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+## Derivative code generation details
+
+The sections above explain derivative code generation, focusing on the transformation rules for function applications (i.e. the `apply` instruction). However, SIL functions are often more complex than just “a single basic block of function applications”:
```
```suggestion
The sections above explain derivative code generation, focusing on the transformation rules for function applications (i.e. the `apply` instruction). However, SIL functions are often more complex than just "a single basic block of function applications":
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+Multiple basic blocks and control flow basic block terminators (`cond_br`, `switch_enum`, etc) require special derivative code generation rules.
+
+> TODO: **Explain the control flow differentiation approach.** The approach is novel and highly worth documenting. [See “Control Flow Differentiation” slides](https://drive.google.com/corp/drive/folders/1y1FeJenXhoz_8xPw0ftVPngJUO_MdmRX) for an overview.
```
```suggestion
> TODO: **Explain the control flow differentiation approach.** The approach is novel and highly worth documenting. [See "Control Flow Differentiation" slides](https://drive.google.com/corp/drive/folders/1y1FeJenXhoz_8xPw0ftVPngJUO_MdmRX) for an overview.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+> TODO: **Explain the control flow differentiation approach.** The approach is novel and highly worth documenting. [See “Control Flow Differentiation” slides](https://drive.google.com/corp/drive/folders/1y1FeJenXhoz_8xPw0ftVPngJUO_MdmRX) for an overview.
+
+- Original functions may have instructions other than `apply`. This includes the entire SIL [instruction set](https://github.com/apple/swift/blob/master/docs/SIL.rst#instruction-set):
```
```suggestion
- Original functions may have instructions other than `apply`. This includes the entire SIL [instruction set](SIL.rst#instruction-set):
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+    - Aggregate operations: `struct`, `tuple`, `enum`.
+    - Projections: `struct_extract`, `tuple_extract`, `struct_element_addr`, `tuple_element_addr`.
+
+Many of these instructions have well-defined corresponding tangent instructions (for the differential) and adjoint instructions (for the pullback). It turns out that supporting transformations of a few common instructions is sufficient for many use cases (e.g. [deep learning APIs](https://github.com/tensorflow/swift-apis)). Here is a short table listing some of these instruction transformation ru ...
```
I'm kinda torn on this.  While the example of the swift-apis is useful, I think it might be better to remove the reference for now.

philipturner replied:
Are you open to adding the reference back in when swift-apis becomes active again? We could also reference @ProfFan's SwiftFusion.

compnerd replied:
I'd rather try to keep the links to the Swift projects.  If swift-apis becomes extremely active due to the community pushing it forward, perhaps that would be reasonable then.

philipturner replied:
Could you clarify which choice you want? Keep as is or replace?

rxwei replied:
I agree with @compnerd.  We should only use links to repos that are part of the Swift open source project (which tensorflow/swift-apis is not) or third-party projects that are actively maintained.

philipturner replied:
@rxwei will hyperlinking to SwiftFusion be sufficient? If not, then we should leave it as is.

compnerd replied:
I think that just removing it without a linked example is fine for the time being.

philipturner replied:
```suggestion
Many of these instructions have well-defined corresponding tangent instructions (for the differential) and adjoint instructions (for the pullback). It turns out that supporting transformations of a few common instructions is sufficient for many use cases. Here is a short table listing some of these instruction transformation rules:
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+### Modules and access levels
+
+Swift organizes code into [modules](https://github.com/apple/swift/blob/master/docs/Modules.rst). Modules enforce access controls on what code can be used outside of them.
```
```suggestion
Swift organizes code into [modules](Modules.rst). Modules enforce access controls on what code can be used outside of them.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+Swift organizes code into [modules](https://github.com/apple/swift/blob/master/docs/Modules.rst). Modules enforce access controls on what code can be used outside of them.
+
+Differentiation is designed with modules and access levels in mind. `@differentiable` declaration attributes act like a “differentiability contract”: declarations marked with the attribute can be differentiated from other modules. This is because the differentiation transform is guaranteed to fill in all SIL differentiability witnesses with derivatives functions. Conversely, declarations not mark ...
```
```suggestion
Differentiation is designed with modules and access levels in mind. `@differentiable` declaration attributes act like a "differentiability contract": declarations marked with the attribute can be differentiated from other modules. This is because the differentiation transform is guaranteed to fill in all SIL differentiability witnesses with derivatives functions. Conversely, declarations not marked with `@differentiable` cannot be differentiated from other modules because they do not have registered derivative functions and are essentially opaque (their bodies may not be exposed).
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+    - If `differentiable_function` has JVP and VJP values, do nothing. It is already canonical.
+    - If `differentiable_function` is missing JVP and VJP:
+        - If the `differentiable_function`’s original function operand is an `differentiable_function_extract [original]` instruction, get the operand of that instruction. Do `differentiable_function_extract [jvp/vjp]` instruction from that operand to get the JVP/VJP. Continue.
+        - Otherwise, get the “original function reference” (`function_ref`, `witness_method`, or `class_method` instruction) underlying the `differentiable_function`’s original function operand.
```
```suggestion
        - Otherwise, get the "original function reference" (`function_ref`, `witness_method`, or `class_method` instruction) underlying the `differentiable_function`’s original function operand.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+    - If `differentiable_function` is missing JVP and VJP:
+        - If the `differentiable_function`’s original function operand is an `differentiable_function_extract [original]` instruction, get the operand of that instruction. Do `differentiable_function_extract [jvp/vjp]` instruction from that operand to get the JVP/VJP. Continue.
+        - Otherwise, get the “original function reference” (`function_ref`, `witness_method`, or `class_method` instruction) underlying the `differentiable_function`’s original function operand.
+        - Look up a `@differentiable` declaration attribute on the “original function reference” whose parameter indices are a minimal superset of the `differentiable_function`’s parameter indices.
```
```suggestion
        - Look up a `@differentiable` declaration attribute on the "original function reference" whose parameter indices are a minimal superset of the `differentiable_function`’s parameter indices.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+            - If the attribute is missing the VJP function, generate a VJP function.
+        - Produce a reference to the JVP/VJP to fill in the `differentiable_function` instruction.
+
+JVP/VJP function generation is explained above. The pseudocode above does not mention how [non-differentiability errors](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md#static-analysis) are handled. If a non-differentiable operation is encountered while processing a SIL differentiability witness or a `differentiable_function` instruction, the transform stops processing  ...
```
```suggestion
JVP/VJP function generation is explained above. The pseudocode above does not mention how [non-differentiability errors](DifferentiableProgramming.md#static-analysis) are handled. If a non-differentiable operation is encountered while processing a SIL differentiability witness or a `differentiable_function` instruction, the transform stops processing the item, continues onto the next item, and finally stops compilation after all items are processed.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1012 @@
+<unknown>:0: note: opaque non-'@differentiable' function is not differentiable
+```
+
+To support reabstraction thunk generation, we must find a way to avoid this “opaque non-@differentiable function” error. One straightforward solution is to change when the `@differentiable` function is formed:
````
```suggestion
To support reabstraction thunk generation, we must find a way to avoid this "opaque non-@differentiable function" error. One straightforward solution is to change when the `@differentiable` function is formed:
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+- Adding a JVP/VJP type calculation special case for reabstraction thunks.
+- Changing the differentiation transform so that reabstraction thunk JVP/VJP callers always construct and pass a `@differentiable` function-typed value.
+
+This approach is implemented in https://github.com/apple/swift/pull/28570. A partially-applied reabstraction thunk derivative matches the derivative type of the reabstracted original function.
```
I'm not sure if we want to reference the pull request itself.  It would be nicer to point to the code, but I don't think that this is a big deal.

philipturner replied:
We have pointed to specific lines of code in other documents (this one and/or the manifesto). That should be possible here.

compnerd replied:
Pointing to the code would be great!

philipturner replied:
There isn't a specific line of code that would work well for this. In addition, the pull request's first comment gives a clear overview of how reabstraction was implemented.

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+This approach is implemented in https://github.com/apple/swift/pull/28570. A partially-applied reabstraction thunk derivative matches the derivative type of the reabstracted original function.
+
+Alternatives:
+- Make reabstraction thunk JVPs/VJPs take a “JVP/VJP” function-typed argument instead of a `@differentiable` function-typed argument.
```
```suggestion
- Make reabstraction thunk JVPs/VJPs take a "JVP/VJP" function-typed argument instead of a `@differentiable` function-typed argument.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+- Make reabstraction thunk JVPs/VJPs take a “JVP/VJP” function-typed argument instead of a `@differentiable` function-typed argument.
+    - This seems more efficient because reabstraction thunk JVPs/VJPs only need to call the original function’s JVP/VJP - the other elements of the `@differentiable` function are not relevant.
+    - It may be possible that the derivative of a reabstraction thunk is simply a reabstraction thunk for the derivative function’s type. If so, this would be a significant simplification: within the reabstraction thunk derivative, we can simply call another reabstraction thunk and avoid other code generation (e.g. reabstraction thunk differential/pullback functions).
+    - `JVPEmitter::visitApplyInst` and `VJPEmitter::visitApplyInst` need special case logic to transform the single `apply` in reabstraction thunks into an `apply` of the “JVP/VJP” function-typed argument.
```
Please update this to reference the correct code in main.

```suggestion
    - `JVPEmitter::visitApplyInst` and `VJPEmitter::visitApplyInst` need special case logic to transform the single `apply` in reabstraction thunks into an `apply` of the "JVP/VJP" function-typed argument.
```

philipturner replied:
Could you provide the links to that code?

rxwei replied:
https://github.com/apple/swift/blob/fecc88702406657ff01d25bb4e1655646e2b3ff7/lib/SILOptimizer/Differentiation/JVPCloner.cpp#L448
https://github.com/apple/swift/blob/fecc88702406657ff01d25bb4e1655646e2b3ff7/lib/SILOptimizer/Differentiation/VJPCloner.cpp#L411

philipturner replied:
```suggestion
    - `JVPCloner::visitApplyInst` and `VJPCloner::visitApplyInst` need special case logic to transform the single `apply` in reabstraction thunks into an `apply` of the "JVP/VJP" function-typed argument.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+Differential-first automatic differentiation is currently a work-in-progress. It is not yet at parity with the existing pullback-first (i.e. reverse-mode) automatic differentiation support.
+
+See [here](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md#differential-operators) for more information about forward-mode differential operators in Swift.
```
```suggestion
See [here](DifferentiableProgramming.md#differential-operators) for more information about forward-mode differential operators in Swift.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+- Linear map function types (`@differentiable(linear)` function types), which are a subtype of `@differentiable` (and infinitely differentiable) function types.
+- Linear map transposition as a first-class operation. This unlocks a correspondence between forward-mode and reverse-mode differentiation: `differentials and pullbacks are transposes of each other`.
+
+See [here](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md#linear-maps) for more information about linear maps. With linear maps and transposition, the differentiation system will change in the following ways:
```
```suggestion
See [here](DifferentiableProgramming.md#linear-maps) for more information about linear maps. With linear maps and transposition, the differentiation system will change in the following ways:
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+- VJP functions will be removed throughout the differentiation system.
+- Pullback function generation will change to use transposition.
+
+## How to contribute? 🤔
+
+[JIRA link to AutoDiff starter bugs](https://bugs.swift.org/issues/?jql=project%20%3D%20TF%20AND%20status%20%3D%20Open%20AND%20component%20in%20(Autodiff)%20AND%20labels%20in%20(StarterBug)). As of the time of this writing, there are no open starter bugs. If you are interested in helping with differentiable programming in Swift, please email [swift@tensorflow.org](mailto:swift@tensorflow.org) and ...
+
```
```suggestion
```

Leaving the JIRA link might be okay as well, if they are still relevant (but that deserves a proper scrub).  The mailto is not appropriate, and this work should encourage contributions as per the regular Swift development practices.

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+## Acknowledgements
+
+Please see [here](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md#acknowledgements) for a list of people who have influenced the design and the implementation of differentiable programming in Swift.
```
```suggestion
Please see [here](DifferentiableProgramming.md#acknowledgements) for a list of people who have influenced the design and the implementation of differentiable programming in Swift.
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+Please see [here](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md#acknowledgements) for a list of people who have influenced the design and the implementation of differentiable programming in Swift.
+
+Some content is borrowed from the [Swift Differentiable Programming Manifesto](https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md) by Richard Wei and [Probabilistic & Differentiable Programming Summit '19 slides](https://twitter.com/rxwei/status/1144688743468527617) - thank you Richard.
```
```suggestion
Some content is borrowed from the [Swift Differentiable Programming Manifesto](DifferentiableProgramming.md) by Richard Wei and [Probabilistic & Differentiable Programming Summit '19 slides](https://twitter.com/rxwei/status/1144688743468527617) - thank you Richard.
```

philipturner commented:
I’ll get to approving these contributions as soon as possible. I constructed this in a private repository off of the main branch, hence the non-relative links.

philipturner committed:
Update DifferentiableProgramming.md

philipturner commented on docs/DifferentiableProgramming.md:
```diff
@@ -698,7 +698,7 @@ differentiation: given an "original function" to differentiate, a derivative
 function with corresponding derivative function applications. The algorithm is
 described in detail in the
-[Swift Differentiable Programming Implementation Overview document](http://bit.ly/swift-autodiff-internals).
+[Swift Differentiable Programming Implementation Overview](./DifferentiableProgrammingImplementation.md).
```
Resolved

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+The VJP function returns a tuple with two elements:
+- The original result of type `U`.
+- A backpropagation function called the “pullback”, which takes the derivatives with respect to the result and returns the derivative with respect to arguments.
```
In the [wikipedia article](https://en.wikipedia.org/wiki/Backpropagation), there is no dash.

compnerd replied:
Ah, weird, seems fine then.

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

I was thinking of removing this since I didn't include it in the table of contents.

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+Here is a description of the main phases in compilation:[†](https://www.swift.org/swift-compiler/)
+- **Parsing**: The parser takes Swift source code and generates an abstract syntax tree (AST) without type information. Warnings and errors are produced for grammatical problems in source code.
+- **Type checking**: The type checker takes the parsed AST and transforms it into a fully type-checked form, performing type inference. Warnings and errors are produced for semantic problems in source code.
+- **SIL generation**: The Swift Intermediate Language (SIL) is a high-level intermediate language for Swift suitable for analysis and optimization. The SIL generation phase lowers the type-checked AST into “raw” SIL.
```
Could you provide a specific link in an additional requested change then?

rxwei replied:
https://github.com/apple/swift/blob/main/docs/SIL.rst#silgen

philipturner replied:
```suggestion
- **SIL generation**: The Swift Intermediate Language (SIL) is a high-level intermediate language for Swift suitable for analysis and optimization. The SIL generation phase lowers the type-checked AST into ["raw" SIL](SIL.rst#silgen).
```

philipturner commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+- **Parsing**: The parser takes Swift source code and generates an abstract syntax tree (AST) without type information. Warnings and errors are produced for grammatical problems in source code.
+- **Type checking**: The type checker takes the parsed AST and transforms it into a fully type-checked form, performing type inference. Warnings and errors are produced for semantic problems in source code.
+- **SIL generation**: The Swift Intermediate Language (SIL) is a high-level intermediate language for Swift suitable for analysis and optimization. The SIL generation phase lowers the type-checked AST into “raw” SIL.
+- **SIL mandatory passes**: The SIL mandatory passes perform analyses and transformations that affect program correctness, including data flow diagnostics and mandatory inlining. The result is “canonical” SIL.
```
I'll approve the quotation change after we get @dan-zheng or @rxwei's feedback.

rxwei replied:
Yeah, "program correctness" is incorrect. It should say something like: 
> The SIL mandatory passes perform analyses that diagnose user errors and lowering to canonical SIL.

rxwei replied:
The commit change (below) was incorrect.
> The SIL mandatory passes perform analyses that diagnose user errors and lower to "canonical" SIL, including data flow diagnostics and mandatory inlining.

It's not analyses that lower to canonical SIL. Rather, the SIL mandatory passes perform analyses and _lowering_. That's why I suggested "lowering". Better suggestions are welcome of course.

philipturner replied:
Could you clarify what you want using the suggested change mechanism?

philipturner replied:
I had a simple misunderstanding of grammatical intention (lowering vs. lower). Could you just show the changes you're asking for using the suggested change mechanism.

rxwei replied:
It's essentially the suggestion in the previous comment (https://github.com/apple/swift/pull/40443#discussion_r764331712).

```suggestion
- **SIL mandatory passes**: The SIL mandatory passes perform analyses (diagnosing user errors) and lowering (down to canonical SIL).
```

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+JVP and VJP functions can be registered using the `@derivative` attribute. JVP functions should return a tuple with labels `(value: ..., differential: ...)`. VJP functions should return a tuple with labels `(value: ..., pullback: ...)`.
+
+For usage examples of the `@derivative` attribute, please see:
+- [The custom differentiation tutorial](https://github.com/tensorflow/swift/blob/master/docs/site/tutorials/custom_differentiation.ipynb), which shows how to use the `@derivative` attribute to register a derivative for `e^x`.
```
I was thinking of eventually turning the tutorial into DocC (see the issue "requesting assistance" on [tensorflow/swift-apis](https://github.com/tensorflow/swift-apis)). However, we can make something else in Markdown for the time being.

philipturner replied:
```suggestion
- [The custom differentiation tutorial](https://github.com/tensorflow/swift/blob/main/docs/site/tutorials/custom_differentiation.ipynb), which shows how to use the `@derivative` attribute to register a derivative for `e^x`.
```

philipturner replied:
Overhauling the tutorial right now would waste more time than any of us can afford. We should leave that task and any changes to the Colab notebooks until we've made a concrete decision of what will become of them.

compnerd replied:
The colab notebooks are no longer useful and we should just drop this if it is not possible to convert.  The colab environment currently does not have the swift integration available.

philipturner replied:
I removed the Jupyter notebook link and merged the remaining bulleted list element into one sentence. Please review the wording.

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1012 @@
+
+### JVP and VJP functions
+
+A JVP (Jacobian-vector products) function is a forward-mode derivative function. A VJP (vector-Jacobian products) function is a reverse-mode derivative function.
```
That seems like a good idea. @dan-zheng could you do this definition?

rxwei replied:
You can add links to the manifesto: https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md#history-of-differentiation-algorithms.

philipturner replied:
```suggestion
A JVP (Jacobian-vector product) function is a [forward-mode](DifferentiableProgramming.md#automatic-differentiation) derivative function. A VJP (vector-Jacobian product) function is a [reverse-mode](DifferentiableProgramming.md#automatic-differentiation) derivative function.
```

philipturner replied:
Waiting for your opinion on changing from plural to singular before committing.

rxwei replied:
LGTM

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented:
I resolved a suggestion somewhere where we changed a link to the `tensorflow` branch to `main`. I just changed a keyword in the url from "tensorflow" to "main" instead of making it a relative link.

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented:
@compnerd I reviewed all of your suggested changes, but some need further discussion.

compnerd commented:
CC: @BradLarson

philipturner commented:
There is also a link that references a "master" branch of apple/swift somewhere in the file.

rxwei commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1005 @@
+Here is the result of activity analysis for this example:
+
+```swift
+// Varied values are in all caps.
+// These values depend on the input (`x`).
+@differentiable
+func f(_ X: Float) -> Float {
+    let SINX = sin(X)
+    let cos3 = cos(3)
+    let RESULT = SINX * cos3
+    print(RESULT)
+    return RESULT
+}
+
+// Useful values are in all caps.
+// These values contribute to the output (`result`).
+@differentiable
+func f(_ X: Float) -> Float {
+    let SINX = sin(X)
+    let COS3 = cos(3) // 3 should be treated as if capitalized.
+    let RESULT = SINX * COS3
+    print(RESULT)
+    return RESULT
+}
+
+// Active values are in all caps.
+// These values are varied and useful, and thus need a derivative.
+@differentiable
+func f(_ X: Float) -> Float {
+    let SINX = sin(X)
+    let cos3 = cos(3)
+    let RESULT = SINX * cos3
+    let void = print(result)
+    return RESULT
+}
````
All caps are a bit hard to read. Can you use a pair of `*` to highlight things?  That wouldn't make it compile but it would convey information more clearly for the purpose of this section.

philipturner replied:
```suggestion
// Varied values are surrounded by asterisks.
// These values depend on the input (`x`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    print(**result**)
    return **result**
}

// Useful values are surrounded by asterisks.
// These values contribute to the output (`result`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let **cos3** = cos(**3**)
    let **result** = **sinx** * **cos3**
    print(**result**)
    return **result**
}

// Active values are surrounded by asterisks.
// These values are varied and useful, and thus need a derivative.
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    let void = print(**result**)
    return **result**
}
```

philipturner replied:
```suggestion
// Varied values are surrounded by asterisks.
// These values depend on the input (`x`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    print(**result**)
    return **result**
}

// Useful values are surrounded by asterisks.
// These values contribute to the output (`result`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let **cos3** = cos(**3**)
    let **result** = **sinx** * **cos3**
    print(**result**)
    return **result**
}

// Active values are surrounded by asterisks.
// These values are varied and useful, and thus need a derivative.
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    let void = print(**result**)
    return **result**
}
```

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Richard Wei <rxrwei@gmail.com>

philipturner committed:
Update DifferentiableProgrammingImplementation.md

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

rxwei commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1005 @@
+  return j(retVal)
+}
+
+bb0(args…):
```
Let's replace all `…` with `...`.

philipturner replied:
Done

rxwei commented on docs/DifferentiableProgrammingImplementation.md:
````diff
@@ -0,0 +1,1005 @@
+
+```swift
+// AD__$sS2fIegnr_S2fIegyd_TR__vjp_src_0_wrt_0
+sil hidden [serializable] [ossa] @AD__$sS2fIegnr_S2fIegyd_TR__vjp_src_0_wrt_0 : $@conven
+tion(thin) (Float, @guaranteed @callee_guaranteed (@in_guaranteed Float) -> @out Float)
````
These should be one line.

philipturner replied:
```suggestion
sil hidden [serializable] [ossa] @AD__$sS2fIegnr_S2fIegyd_TR__vjp_src_0_wrt_0 : $@convention(thin) (Float, @guaranteed @callee_guaranteed (@in_guaranteed Float) -> @out Float)
```

rxwei commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1005 @@
+-> (Float, @owned @callee_guaranteed (Float) -> Float) {
+// %0                                             // user: %3
+// %1                                             // user: %5
+bb0(%0 : $Float, %1 : @guaranteed $@callee_guaranteed (@in_guaranteed Float) -> @out Flo
+at):
```
One line here

philipturner replied:
```suggestion
bb0(%0 : $Float, %1 : @guaranteed $@callee_guaranteed (@in_guaranteed Float) -> @out Float):
```

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

rxwei commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1005 @@
+// fills in JVP/VJP functions.
+
+sil_differentiability_witness hidden [parameters 0] [results 0] @$s3fooAAyS2fF : $@convention(thin) (Float) -> Float {
+  jvp: @AD__$s3fooAAyS2fF__jvp_src_0_wrt_0 : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float)
+  vjp: @AD__$s3fooAAyS2fF__vjp_src_0_wrt_0 : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float)
```
Note that these symbols are no longer mangled this way.  We should start a separate patch and update these function names. You can write a simple program with `foo(_:)`, compile with `swiftc -emit-sil` and find its JVP and VJP's mangled names in the corresponding differentiability witness.

philipturner replied:
I will leave that up to you, as I am not that familiar with the compiler yet.

philipturner replied:
@rxwei Also, I need to import the Differentiation package to make the code compile. This means I can only compile it in Xcode, not in the command line (without more knowledge of the compiler). If you have a development toolchain installed, then you might have better luck with it.

rxwei replied:
`_Differentiation` is always included in development toolchains. You just need to import `_Differentiation`. That works with command line as well.

compnerd replied:
@rxwei awesome!  The decoration scheme was something that really bothered me.  Is the current scheme documented?

rxwei replied:
Everything's documented in https://github.com/apple/swift/blob/main/docs/ABI/Mangling.rst (search for "autodiff" and "differentiable").

philipturner replied:
<img width="1257" alt="Screen Shot 2021-12-07 at 8 06 00 PM" src="https://user-images.githubusercontent.com/71743241/145130139-f36af3e6-5680-4648-a59d-d9de3bd67c0c.png">

I don't have a development toolchain configured. If it's not too much to ask, could @compnerd or @rxwei fix the names?

rxwei replied:
You don't need to fix the names to merge this PR.  We can do it later.

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Richard Wei <rxrwei@gmail.com>

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Richard Wei <rxrwei@gmail.com>

philipturner committed:
Update DifferentiableProgrammingImplementation.md

philipturner committed:
Merge pull request #1 from philipturner/philipturner-patch-4-pt-1

Update DifferentiableProgrammingImplementation.md

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Update DifferentiableProgrammingImplementation.md

philipturner committed:
Update DifferentiableProgrammingImplementation.md

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1003 @@
+# Swift Differentiable Programming Implementation Overview
+
+- Authors: [Dan Zheng](https://github.com/dan-zheng), [Bart Chrzaszcz](https://github.com/bartchr808), [Philip Turner](https://github.com/philipturner)
```
Have you made extensive changes to the document that from the original document?   I'm asking as you are marking yourself as an author, and I'd like to understand what the difference is from the original document being converted.

philipturner replied:
Look at my PR to swift-apis (hyperlinked in the first comment to this PR). I outlined 5 changes I made to the original document. @compnerd please look over the other 4 and validate them.

rxwei replied:
We really appreciate your contribution to this document, but the authorship in this document was meant to reflect substantial _technical_ contribution from years of design and development in addition to just writing.  As a result, I believe it's more appropriate to keep the original author list.  Be assured that your contribution will still be acknowledged, regardless of being listed as an author in the document.  Thanks!

philipturner replied:
```suggestion
- Authors: [Dan Zheng](https://github.com/dan-zheng), [Bart Chrzaszcz](https://github.com/bartchr808)
```

compnerd commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1003 @@
+
+### Visiting the original function to create a JVP
+
+In the Swift compiler, we have a class called [`JVPEmitter`](https://github.com/apple/swift/blob/main/lib/SILOptimizer/Mandatory/Differentiation.cpp) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate the JVP  ...
```
```suggestion
In the Swift compiler, we have a class called [`JVPCloner`](https://github.com/apple/swift/blob/main/include/swift/SILOptimizer/Differentiation/JVPCloner.h) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate the JVP function. Taking a look at the `JVPCloner` class, there are methods like `visitApplyInst`, `visitStructExtractInst`, etc. Each of these methods visit an instruction that is important in the generation of the JVP and differential, and emits a newly mapped version. We handle each type of instruction differently, explained below.
```

This expects a fair understanding of C++ as I assume that the reader will understand the PIMPL pattern and how it applies to this.

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Please review - eliminate bulleted list

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner committed:
Make two links relative - please review

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

philipturner commented:
One more pending requested change before this can merge!

rxwei commented on docs/DifferentiableProgrammingImplementation.md:
```diff
@@ -0,0 +1,1000 @@
+
+JVP and VJP functions can be registered using the `@derivative` attribute. JVP functions should return a tuple with labels `(value: ..., differential: ...)`. VJP functions should return a tuple with labels `(value: ..., pullback: ...)`.
+
+For usage examples of the `@derivative` attribute, please see [this mailing list response](https://groups.google.com/a/tensorflow.org/d/msg/swift/k62Uwnc-EYc/Cncb7mxtCAAJ), which shows how to use the `@derivative` attribute to register a derivative for `ax^2 + bx + c`.
```
I don't think we should refer to mailing list responses to show examples.  Usage of `@derivative` is well-documented in https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#derivative-attribute and we should link to that instead.

philipturner replied:
Could you provide a suggested change of exactly how you want the new sentence worded?

rxwei replied:
For usage examples of the `@derivative` attribute, please see https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#derivative-attribute.

philipturner replied:
```suggestion
For usage examples of the `@derivative` attribute, refer to the [`@derivative` attribute section](https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#derivative-attribute) in the proposal.
```

philipturner replied:
Is this okay now?

rxwei replied:
Yes. Thanks.

philipturner committed:
Update docs/DifferentiableProgrammingImplementation.md

Co-authored-by: Richard Wei <rxrwei@gmail.com>

philipturner commented:
This should be ready for merging now

rxwei commented:
Thanks for addressing the comments.  I'll give others (@dan-zheng, @BradLarson) some time to have a look before approving.

rxwei commented:
@swift-ci please smoke test and merge

philipturner commented:
Finally!

dan-zheng commented:
Sorry for the late review. The ported document looks good, thank you @philipturner!
